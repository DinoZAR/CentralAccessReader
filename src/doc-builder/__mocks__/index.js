const DocBuilder = require.requireActual('../index').default;

const TRACKED_CALLS = new Set([
  'paragraph',
  'text',
  'math',
  'image',
]);

const UNTOUCHED_CALLS = new Set([
  'build'
]);

let __calls = [];

function MockBuilder(tmpFolder) {
  DocBuilder.call(this, tmpFolder);
}
MockBuilder.prototype = Object.entries(DocBuilder.prototype).reduce((accum, [funcName, func]) => {
  if (TRACKED_CALLS.has(funcName)) {
    return {
      ...accum,
      [funcName]: function(...params) {
        __calls.push({
          func: funcName,
          params,
        });
      }
    };
  } else if (UNTOUCHED_CALLS.has(funcName)) {
    return {
      ...accum,
      [funcName]: func
    };
  }
  return {
    ...accum,
    [funcName]: () => {}
  };
}, {});

export function calls() {
  return __calls;
}

export function resetCalls() {
  __calls = [];
}

export default MockBuilder;