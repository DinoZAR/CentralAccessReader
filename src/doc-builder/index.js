import fs from 'fs';
import path from 'path';

/**
 * Builds up a document and saves the chunks to the temp directory.
 *
 * @param {string} tmpFolder - Folder that was allocated from the importer
 */
function DocBuilder(tmpFolder) {
  this.tmpFolder = tmpFolder;
  this.idCounter = 0;
  this.chunkCounter = 0;
  this.content = [];

  this.paragraphInfo = null;
  this.paragraphContent = [];
}

DocBuilder.prototype.__newID = function() {
  this.idCounter += 1;
  return this.idCounter;
};

DocBuilder.prototype.__newChunkID = function() {
  const prevID = this.chunkCounter;
  this.chunkCounter += 1;
  return prevID;
};

DocBuilder.prototype.__writeParagraph = function() {
  this.content.push({
    ...this.paragraphInfo,
    content: this.paragraphContent,
  });
  this.paragraphInfo = null;
  this.paragraphContent = [];
};

DocBuilder.prototype.__writeChunk = function() {
  const id = this.__newChunkID();
  fs.writeFile(
    path.join(this.tmpFolder, '__chunks', `${id}.json`),
    JSON.stringify(this.content),
    (err) => {
      if (err) throw err;
    }
  );
};

DocBuilder.prototype.__done = function() {
  this.__writeParagraph();
  this.__writeChunk();
};

DocBuilder.prototype.build = function(buildFunc) {
  buildFunc(this);
  this.__done();
};

DocBuilder.prototype.paragraph = function() {
  this.paragraphInfo = {
    id: this.__newID(),
    type: 'paragraph',
  };
};

DocBuilder.prototype.text = function({ content }) {
  this.paragraphContent.push({
    id: this.__newID(),
    type: 'text',
    content,
  });
};

DocBuilder.prototype.math = function({ mathml }) {
  this.paragraphContent.push({
    id: this.__newID(),
    type: 'math',
    mathml,
  });
};

DocBuilder.prototype.image = function({ url }) {
  this.paragraphContent.push({
    id: this.__newID(),
    type: 'image',
    url,
  });
};

export default DocBuilder;