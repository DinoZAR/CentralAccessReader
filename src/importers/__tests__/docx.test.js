import path from 'path';
import importDocx from '../docx';
import { calls, resetCalls } from '../../doc-builder';

jest.mock('../../doc-builder');

describe('importDocx()', () => {
  it('parses a basic doc with text in it', () => {
    resetCalls();
    return importDocx(path.join(__dirname, './test_files/basic-text.docx'), '/tmp').then(() => {
      expect(calls()).toMatchSnapshot();
    });
  });

  it('parses images and text', () => {
    resetCalls();
    return importDocx(path.join(__dirname, './test_files/text-with-images.docx'), '/tmp').then(() => {
      expect(calls()).toMatchSnapshot();
    });
  });
});