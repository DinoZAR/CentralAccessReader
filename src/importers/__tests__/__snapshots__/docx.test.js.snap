// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importDocx() parses a basic doc with text in it 1`] = `
Array [
  Object {
    "func": "paragraph",
    "params": Array [],
  },
  Object {
    "func": "text",
    "params": Array [
      Object {
        "content": "A very basic word doc.",
      },
    ],
  },
  Object {
    "func": "paragraph",
    "params": Array [],
  },
  Object {
    "func": "text",
    "params": Array [
      Object {
        "content": "Look, another paragraph!",
      },
    ],
  },
]
`;

exports[`importDocx() parses images and text 1`] = `
Array [
  Object {
    "func": "paragraph",
    "params": Array [],
  },
  Object {
    "func": "text",
    "params": Array [
      Object {
        "content": "Some text",
      },
    ],
  },
  Object {
    "func": "image",
    "params": Array [
      Object {
        "deferred": true,
        "deferredData": Object {
          "altText": "",
          "zipURL": "media/image1.jpg",
        },
      },
    ],
  },
  Object {
    "func": "paragraph",
    "params": Array [],
  },
  Object {
    "func": "paragraph",
    "params": Array [],
  },
  Object {
    "func": "image",
    "params": Array [
      Object {
        "deferred": true,
        "deferredData": Object {
          "altText": "",
          "zipURL": "media/image2.jpg",
        },
      },
    ],
  },
]
`;
